src/backends/wasm_js.rs

```diff
@@ -7,32 +7,48 @@ pub use crate::util::{inner_u32, inner_u64};
 #[cfg(not(all(target_arch = "wasm32", any(target_os = "unknown", target_os = "none"))))]
 compile_error!("`wasm_js` backend can be enabled only for OS-less WASM targets!");
 
-use js_sys::{global, Uint8Array};
-use wasm_bindgen::{prelude::wasm_bindgen, JsCast, JsValue};
+#[cfg(target_feature = "atomics")]
+use js_sys::Uint8Array;
+use wasm_bindgen::{prelude::wasm_bindgen, JsValue};
 
-// Size of our temporary Uint8Array buffer used with WebCrypto methods
-// Maximum is 65536 bytes see https://developer.mozilla.org/en-US/docs/Web/API/Crypto/getRandomValues
-const CRYPTO_BUFFER_SIZE: u16 = 256;
+// Maximum buffer size allowed in `Crypto.getRandomValuesSize` is 65536 bytes.
+// See https://developer.mozilla.org/en-US/docs/Web/API/Crypto/getRandomValues
+const MAX_BUFFER_SIZE: u16 = 256;
 
 pub fn fill_inner(dest: &mut [MaybeUninit<u8>]) -> Result<(), Error> {
-    let global: Global = global().unchecked_into();
-    let crypto = global.crypto();
+    CRYPTO.with(|crypto| {
+        let crypto = crypto.as_ref().ok_or(Error::WEB_CRYPTO)?;
+        inner(crypto, dest)
+    })
+}
 
-    if !crypto.is_object() {
-        return Err(Error::WEB_CRYPTO);
+#[cfg(not(target_feature = "atomics"))]
+fn inner(crypto: &Crypto, dest: &mut [MaybeUninit<u8>]) -> Result<(), Error> {
+    for chunk in dest.chunks_mut(MAX_BUFFER_SIZE.into()) {
+        if crypto.get_random_values(chunk).is_err() {
+            return Err(Error::WEB_GET_RANDOM_VALUES);
+        }
     }
+    Ok(())
+}
 
+#[cfg(target_feature = "atomics")]
+fn inner(crypto: &Crypto, dest: &mut [MaybeUninit<u8>]) -> Result<(), Error> {
     // getRandomValues does not work with all types of WASM memory,
     // so we initially write to browser memory to avoid exceptions.
-    let buf = Uint8Array::new_with_length(CRYPTO_BUFFER_SIZE.into());
-    for chunk in dest.chunks_mut(CRYPTO_BUFFER_SIZE.into()) {
+    let buf = Uint8Array::new_with_length(MAX_BUFFER_SIZE.into());
+    for chunk in dest.chunks_mut(MAX_BUFFER_SIZE.into()) {
         let chunk_len: u32 = chunk
             .len()
             .try_into()
-            .expect("chunk length is bounded by CRYPTO_BUFFER_SIZE");
+            .expect("chunk length is bounded by MAX_BUFFER_SIZE");
         // The chunk can be smaller than buf's length, so we call to
         // JS to create a smaller view of buf without allocation.
-        let sub_buf = buf.subarray(0, chunk_len);
+        let sub_buf = if chunk_len == u32::from(MAX_BUFFER_SIZE) {
+            buf.clone()
```

Comment:
Does this mean that we create a new allocation here instead of reusing the existing one?

Reply:
No, cloning the buffer just clones the reference, not actually the whole buffer.
This is just here to save the FFI call to `subarray()`.

I did it like this because our MSRV of v1.63 doesn't support [propagating a temporary lifetime extension into if](https://github.com/rust-lang/rust/pull/121346), but now that we have gated this behind `target_feature = "atomics"` this isn't MSRV bound anymore.

Reply:
Done.